### PR TITLE
Caching related actors

### PR DESCRIPTION
**Pull request by wintondeshong**
Addition of caching related actors for key generation and data serialization.

- [x] Related GitHub issue(s) linked in PR description
- [x] Destination branch merged, built and tested with your changes
- [x] Code formatted and follows best practices and patterns
- [x] Code builds cleanly (no _additional_ warnings or errors)
- [x] Manually tested
- [x] Automated tests are passing
- [x] No _decreases_ in automated test coverage
- [x] Documentation updated (readme, docs, comments, etc.)
- [x] Localization: No hard-coded error messages in code files (_minimally_ in string constants)

### CODE REVIEW

**Review by Stefanie899 — APPROVED** [review]
Looks good to me!

File: test/AndcultureCode.CSharp.Core.Tests/Constants/CacheOptionsTest.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+using AndcultureCode.CSharp.Core.Models;
+using AndcultureCode.CSharp.Core.Tests.Stubs;
+using AndcultureCode.CSharp.Core.Tests.Unit;
+using AndcultureCode.CSharp.Testing;
+using AndcultureCode.CSharp.Testing.Tests;
+using Shouldly;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace AndcultureCode.CSharp.Core.Tests.Models
```
Comment (brandongregoryscott) [inline test/AndcultureCode.CSharp.Core.Tests/Constants/CacheOptionsTest.cs:12]: Namespace is off here

File: test/AndcultureCode.CSharp.Core.Tests/Extensions/IDistributedCacheExtensionsTest.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            var sut = Mock.Of<IDistributedCache>();
+
+            // Act
+            var result = sut.Set<string>(key: key, value: Random.String(), options: null);
+
+            // Assert
+            result.ShouldBeNull();
+        }
+
+        public static IEnumerable<object[]> GetSetCacheData(int numTests) => new List<object[]>
```
Comment (brandongregoryscott) [inline test/AndcultureCode.CSharp.Core.Tests/Extensions/IDistributedCacheExtensionsTest.cs:85]: Recommend adding a comment above this function explaining its purpose & moving it to a setup region (whether it's at the top of the file or just the tests for `Set<T>`)

**Review by brandongregoryscott — APPROVED** [review]
Left some minor feedback 👍 Looks good though!

File: test/AndcultureCode.CSharp.Core.Tests/Extensions/IDistributedCacheExtensionsTest.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+            // Act
+            var result = sut.Set<string>(key: key, value: Random.String(), options: null);
+
+            // Assert
+            result.ShouldBeNull();
+        }
+
+        [Theory]
+        [MemberData(nameof(SetupSetData))]
```
Comment (HeyKos) [inline test/AndcultureCode.CSharp.Core.Tests/Extensions/IDistributedCacheExtensionsTest.cs:96]: Praise: Excellent use of `MemberData` here.

File: test/AndcultureCode.CSharp.Core.Tests/Extensions/IDistributedCacheExtensionsTest.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            // Arrange
+            var key = Random.String();
+            var sut = new Mock<IDistributedCache>();
+            var expected = sut.Object.Serialize(value);
+
+            // Act
+            var result = sut.Object.Set<object>(key: key, value: value, options: null);
+
+            // Assert
+            result.ShouldBe(expected);
```
Comment (HeyKos) [inline test/AndcultureCode.CSharp.Core.Tests/Extensions/IDistributedCacheExtensionsTest.cs:124]: Praise: I like that you broke out your asserts into separate tests. It makes it clearer what aspects of the SUT you want to verify.

File: test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        public CacheKeyAttributeTest(ITestOutputHelper output) : base(output) { }
+
+        #endregion Setup
+
+        #region GetKey
+
+        [Fact]
+        public void GetKey_When_Id_Set_Returns_Key_Containing_Type()
+        {
+            CacheKeys.GetKey<CultureStub>(id: Random.Long()).ShouldContain(typeof(CultureStub).Name);
```
Comment (HeyKos) [inline test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs:25]: Nitpick: Consider adding `// Arrange, Act, Assert` here. Not sure how much I care though.

Comment (HeyKos) [inline test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs:25]: Suggestion: As with earlier comment consider changing  `typeof(CultureStub).Name` to `nameof(CultureStub)`.

File: test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            // Arrange
+            var expected = Random.Long();
+
+            // Act & Assert
+            CacheKeys.GetKey<CultureStub>(id: expected).ShouldContain(expected.ToString());
+        }
+
+        [Fact]
+        public void GetKey_When_Include_Properties_Set_Returns_Key_Containing_Sorted_Properties()
```
Comment (HeyKos) [inline test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs:39]: Discussion: Seeing this test now gives me more insight into the intent of the ordering by property name. Without understanding the details, is there a chance that the scenario I mentioned on line 80 in `CacheKey.cs` on could occur?

File: test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            // Arrange
+            var delimiter = "-";
+            var includeProperties = new List<Expression<Func<CultureStub, object>>>
+            {
+                e => e.IsDefault, // <--- intentionally out of sort order
```
Comment (HeyKos) [inline test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs:45]: Praise: Love the clarity on the important aspect of the arrangement.

**Review by HeyKos — COMMENTED** [review]
Awesome job with this Winton!

File: src/AndcultureCode.CSharp.Core/Extensions/ObjectExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    /// </summary>
+    public static class ObjectExtensions
+    {
+        /// <summary>
+        /// Converts supplied byte array to requested type
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        public static T FromByteArray<T>(this byte[] byteArray) where T : class
+        {
+            if (byteArray == null)
```
[the remark this answers is not in the document]
Comment (wintondeshong) [reply src/AndcultureCode.CSharp.Core/Extensions/ObjectExtensions.cs:18]: Yeah, objects that are possibly `null` can still have extension methods on them.

File: src/AndcultureCode.CSharp.Core/Utilities/Caching/CacheKeys.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            if (includeProperties.IsNullOrEmpty())
+            {
+                return key;
+            }
+
+            var names = includeProperties
+                .Select(GetPropertyName)
+                .Distinct()
+                .OrderBy(p => p)
```
[the remark this answers is not in the document]
Comment (wintondeshong) [reply src/AndcultureCode.CSharp.Core/Utilities/Caching/CacheKeys.cs:80]: Yeah, this is intentional to try our best to have cache key hits.

File: src/AndcultureCode.CSharp.Core/Constants/CacheOptions.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+        #region Methods
+
+        /// <summary>
+        /// Configure options for X minutes sliding scale expiration
+        /// </summary>
+        /// <param name="minutes"></param>
+        /// <returns></returns>
+        public static DistributedCacheEntryOptions Minutes(double minutes)
+            => new DistributedCacheEntryOptions { SlidingExpiration = TimeSpan.FromMinutes(minutes) };
```
[the remark this answers is not in the document]
Comment (wintondeshong) [reply src/AndcultureCode.CSharp.Core/Constants/CacheOptions.cs:28]: good catch!

File: src/AndcultureCode.CSharp.Core/Utilities/Caching/CacheKeys.cs [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        /// Construct standardized cache keys for entities with optional include properties
+        /// </summary>
+        /// <param name="id"></param>
+        /// <param name="delimiter"></param>
+        /// <param name="includeProperties"></param>
+        public static string GetKey<T, TProperties>(
+            long id,
+            string delimiter = DEFAULT_DELIMITER,
+            List<Expression<Func<TProperties, object>>> includeProperties = null
+        ) => AppendIncludeProperties($"{typeof(T).Name}{delimiter}{id}", delimiter, includeProperties);
```
[the remark this answers is not in the document]
Comment (wintondeshong) [reply src/AndcultureCode.CSharp.Core/Utilities/Caching/CacheKeys.cs:63]: Being `nameof` is compiled into a constant, it results in undesired behavior. I infact do want the reflection version of `typeof(T).Name`. After changing to `nameof(T)` my test failed and returned...

```
expected
"CultureStub"
    but was actually
"T--2424723088006234127"
``

File: test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs (hunk 5 above)
Comment (wintondeshong) [reply test/AndcultureCode.CSharp.Core.Tests/Utilities/Caching/CacheKeyTest.cs:25]: in this case, nameof should work being i don't need it to be dynamic at runtime

**Comment by wintondeshong** [comment]
@HeyKos I submitted another PR accounting for your feedback https://github.com/AndcultureCode/AndcultureCode.CSharp.Core/pull/69 Thanks!

**Pull request merged**